Enable proxying through yourself

dborkan opened the pull request:
Fixes https://github.com/uProxy/uproxy/issues/688

We now allow proxying through other instances of yourself, which works in both Google and Facebook.

Some additional changes:
* Other instances of yourself are automatically permissioned for access
* We don't display you as a user in the buddylist unless you have another instance (i.e. multiple instances).  This is to prevent brand new users from signing in with Facebook and just seeing themselves on the buddylist
* We don't display desktop notifications for permission changes for other instances of yourself (to prevent "X has granted you access" from automatically appearing as soon as you sign on in 2 places with the same user)
* We only show the sharing enabled icon and tooltip if you are sharing with someone other than yourself
* Removed some bad tests

Tested in Chrome, Firefox, and with updated grunt tests

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1569)
<!-- Reviewable:end -->

willscott commented:
Very excited to see this :smile:

jpevarnek commented:
A few small comments, mostly nit-picking

---

Review status: 9 of 11 files reviewed, 9 unresolved discussions, some commit checks failed.
Reviewed files:
- src/generic_core/remote-instance.spec.ts @ r1
- src/generic_core/remote-user.spec.ts @ r1
- src/generic_core/remote-user.ts @ r1
- src/generic_core/social.spec.ts @ r1
- src/generic_core/social.ts @ r1
- src/generic_core/uproxy_core.spec.ts @ r1
- src/generic_ui/polymer/root.html @ r1
- src/generic_ui/polymer/root.ts @ r1
- src/generic_ui/scripts/user.spec.ts @ r1

---

<sup>**[src/generic_core/consent.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwLiLGkqpT8dW6L_aV)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/consent.ts#L28)):</sup>
This name feels a bit offer (it makes me think of "Initialize consent" instead of "Initial Consent")

---

<sup>**[src/generic_core/remote-user.spec.ts, line 396 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwNO7J1no2nE7QoEcI)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/remote-user.spec.ts#L396)):</sup>
Testing the ignoring parameters seems a bit outside the scope of the test description, maybe move those into separate tests?

---

<sup>**[src/generic_core/remote-user.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwNWTIYapIeaa0QkXn)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/remote-user.ts#L100)):</sup>
Is there a problem with using `===` here?

---

<sup>**[src/generic_core/remote-user.ts, line 359 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwNzyPd8A0_y_IWX4F)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/remote-user.ts#L359)):</sup>
It seems like this might read a bit more clearly as `if (allInstanceIds.length === 0 && (!this.network.areAllContactsUproxy() || this.userId === this.network.myInstance.userId))`.  I'm not terribly attached to splitting that out, but it feels weird having one cases handled at two levels while the other case is handled in the if statement with an implied second condition from the else.

---

<sup>**[src/generic_core/social.ts, line 222 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwP3Qp08p9EYUnBGYq)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/social.ts#L222)):</sup>
Update comment

---

<sup>**[src/generic_core/social.ts, line 367 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwPBblh6njansyQzuw)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/social.ts#L367)):</sup>
Is there any way to deprecate this message and have the UI get the data from the normal profile message?

---

<sup>**[src/generic_ui/polymer/root.ts, line 218 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwPh3HgZD4IJG7gZhK)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_ui/polymer/root.ts#L218)):</sup>
Can you add a small comment on why you have the observer here?

---

<sup>**[src/generic_ui/polymer/root.ts, line 222 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwQ4P1nsKqmi9E4Upu)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_ui/polymer/root.ts#L222)):</sup>
`!==`

---

<sup>**[src/generic_ui/scripts/user.ts, line 77 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwPqkI0EIZi_i1axaE)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_ui/scripts/user.ts#L77)):</sup>
`!==`?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1569)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Review status: all files reviewed, 9 unresolved discussions, some commit checks failed.
Reviewed files:
- src/generic_core/consent.ts @ r1
- src/generic_ui/scripts/user.ts @ r1

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1569)
<!-- Sent from Reviewable.io -->

dborkan commented:
Just made those changes, PTAL

---

Review status: 5 of 11 files reviewed, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/consent.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwLiLGkqpT8dW6L_aV)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/consent.ts#L28)):</sup>
Changed to initialConsent

---

<sup>**[src/generic_core/remote-user.spec.ts, line 396 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwNO7J1no2nE7QoEcI)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/remote-user.spec.ts#L396)):</sup>
I've removed the ignoring bits from these test

---

<sup>**[src/generic_core/social.ts, line 367 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1569#-JpwPBblh6njansyQzuw)** ([raw file](https://github.com/uproxy/uproxy/blob/12b8c5efeee07bc914ddb015476805f25d84038f/src/generic_core/social.ts#L367)):</sup>
Discussed with Jonathan in person, ignoring this for now

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1569)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
:thumbsup:

---

Review status: all files reviewed, all discussions resolved, some commit checks failed.
Reviewed files:
- src/generic_core/consent.ts @ r2
- src/generic_core/remote-user.spec.ts @ r2
- src/generic_core/remote-user.ts @ r2
- src/generic_core/social.ts @ r2
- src/generic_ui/polymer/root.ts @ r2
- src/generic_ui/scripts/user.ts @ r2

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1569)
<!-- Sent from Reviewable.io -->
